go/vt/sqlparser/ast_funcs.go

```diff
@@ -1447,13 +1452,14 @@ const (
 
 // handleUnaryMinus handles the case when a unary minus operator is seen in the parser. It takes 1 argument which is the expr to which the unary minus has been added to.
 func handleUnaryMinus(expr Expr) Expr {
-	if num, ok := expr.(*Literal); ok && (num.Type == IntVal || num.Type == FloatVal) {
+	if num, ok := expr.(*Literal); ok && (num.Type == IntVal || num.Type == FloatVal || num.Type == DecimalVal) {
 		// Handle double negative
 		if num.Val[0] == '-' {
 			num.Val = num.Val[1:]
 			return num
 		}
-		return NewIntLiteral("-" + num.Val)
+		num.Val = "-" + num.Val
+		return num
```

Comment:
cc @systay  -- this bug has been present for ages. Whenever we saw a literal float preceded by unary minus, this hardcoded call to `NewIntLiteral` would emit it with an integer type, even though the actual value of the bytes for the literal would still be a floating point number.

Reply:
:facepalm: